### PR TITLE
Reenable HardwareBuffer backed Android Platform Views on SDK >= 29

### PR DESCRIPTION
**Pull request by johnmccutchan**
- Fix a bug in the SDK < 33 ImageReader construction code path.
- Fix a bug that resulted in references to Images produced by a closed ImageReader.
- Fix an order of operations bug in ImageReaderPlatformViewRenderTarget release/finalizer code path.
- Enable HardwareBuffer backed Android Platform Views on SDK >= 29

Manually tested on device rotating and shutting down the app.

### CODE REVIEW

**Comment by jason-simmons** [comment]
Tried this PR on a Pixel 3.

If I run the https://github.com/flutter/flutter/tree/master/dev/integration_tests/abstract_method_smoke_test test and rotate the screen, then the `PlatformViewAndroidJNIImpl::ImageGetHardwareBuffer` JNI call gets an uncaught exception.

File: shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java [hunk 1]
```diff
@@ -68,29 +68,31 @@ protected ImageReader createImageReader33() {
...
       return createImageReader29();
-    } else {
-      return createImageReader26();
     }
+    throw new UnsupportedOperationException(
+        "ImageReaderPlatformViewRenderTarget requires API version 29+");
   }
 
   public ImageReaderPlatformViewRenderTarget(ImageTextureEntry textureEntry) {
     if (Build.VERSION.SDK_INT < 26) {
```
Comment (navaronbracke) [inline shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java:93]: This should be 29, instead of 26.

Comment (johnmccutchan) [reply shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java:93]: Good catch. Done.

**Comment by johnmccutchan** [comment]
@jason-simmons This is ready for review.

File: shell/platform/android/hardware_buffer_external_texture.cc [hunk 2]
```diff
@@ -39,7 +39,8 @@ void HardwareBufferExternalTexture::Paint(PaintContext& context,
         flutter::DlCanvas::SrcRectConstraint::kStrict  // enforce edges
     );
   } else {
-    FML_LOG(WARNING) << "No DlImage available.";
+    FML_LOG(WARNING)
+        << "No DlImage available for HardwareBufferExternalTexture to paint.";
```
Comment (dnfield) [inline shell/platform/android/hardware_buffer_external_texture.cc:43]: What's the developer action to take here?

If there is none, maybe this should be a DLOG/DCHECK?

File: shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java [hunk 3]
```diff
@@ -55,9 +48,12 @@ protected ImageReader createImageReader33() {
...
     // Use PRIVATE image format so that we can support video decoding.
-    // TODO(johnmccutchan): Should we always use PRIVATE here? It may impact our ability
-    // to read back texture data. If we don't always want to use it, how do we decide when
-    // to use it or not? Perhaps PlatformViews can indicate if they may contain DRM'd content.
+    // TODO(johnmccutchan): Should we always use PRIVATE here? It may impact our
+    // ability
+    // to read back texture data. If we don't always want to use it, how do we
+    // decide when
+    // to use it or not? Perhaps PlatformViews can indicate if they may contain
+    // DRM'd content.
```
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java:56]: nit: formatting
nit: please link an issue here explaining what has to be done (even if just what investigation is left to do or what resources someone interested in this should read about)

File: shell/platform/android/surface_texture_external_texture.cc [hunk 4]
```diff
@@ -65,7 +65,8 @@ void SurfaceTextureExternalTexture::Paint(PaintContext& context,
         flutter::DlCanvas::SrcRectConstraint::kStrict  // enforce edges
     );
   } else {
-    FML_LOG(WARNING) << "No DlImage available.";
+    FML_LOG(WARNING)
+        << "No DlImage available for SurfaceTextureExternalTexture to paint.";
```
Comment (dnfield) [inline shell/platform/android/surface_texture_external_texture.cc:69]: Ditto on log message.

**Review by dnfield — COMMENTED** [review]
Few nits

File: shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java (hunk 3 above)
Comment (johnmccutchan) [reply shell/platform/android/io/flutter/plugin/platform/ImageReaderPlatformViewRenderTarget.java:56]: Fixed added more context to the TODO.

File: shell/platform/android/hardware_buffer_external_texture.cc (hunk 2 above)
Comment (johnmccutchan) [reply shell/platform/android/hardware_buffer_external_texture.cc:43]: This is useful information for debugging the engine and I'd like to keep it in place for now until this code is more solid.

File: shell/platform/android/surface_texture_external_texture.cc (hunk 4 above)
Comment (johnmccutchan) [reply shell/platform/android/surface_texture_external_texture.cc:69]: Same here.

**Comment by jason-simmons** [comment]
LGTM

**Review by jonahwilliams — APPROVED** [review]
RSLGTM

**Pull request merged**